### PR TITLE
Incorporated some comments - Net Secure Xplat Implementation

### PR DESCRIPTION
**Pull request by rajansingh10**
Incorporated some of the comments.
Sending it as a new PR as original PR will be closed, once we merge into dev_sslstream to unblock partners.
Tracking all issues via #3362 . (I am yet to updated the issue with complete list).
Updated all TODO's in the code to refer this issue.
This also includes fixes from Vijay's dummy PR https://github.com/dotnet/corefx/pull/3365.

### CODE REVIEW

**Comment by rajansingh10** [comment]
cc: @stephentoub, @bartonjs, @davidsh, @CIPop, @SidharthNabar, @pgavlin , @vijaykota,

**Comment by rajansingh10** [comment]
@dotnet-bot test this please

**Comment by bartonjs** [comment]
I liked how @vijaykota had things broken up into cleaner commits (and it let me cherry pick them), so some parts of this are being included with #3386.

**Comment by rajansingh10** [comment]
@dotnet-bot test this please

**Comment by rajansingh10** [comment]
@dotnet-bot test this please

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs [hunk 1]
```diff
@@ -212,25 +225,27 @@ private long GetOptions(SslProtocols protocols)
...
+                if (outputPtr != IntPtr.Zero)
+                {
+                     Marshal.FreeHGlobal(outputPtr);
+                     outputPtr = IntPtr.Zero;
+                }
 
                 return done ? SecurityStatus.OK : SecurityStatus.ContinueNeeded;
             }
-            catch
+            catch(Exception Ex)
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: Naming: ex instead of Ex (locals should be camelCased).

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs [hunk 2]
```diff
@@ -212,25 +225,27 @@ private long GetOptions(SslProtocols protocols)
...
             {
-                return SecurityStatus.InternalError;
+                Debug.Fail("Exception Caught. StackTrace - " +Ex.StackTrace);
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: Why only print the StackTrace value?  Why not just `ex` (implicit ToString)?

File: src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs [hunk 3]
```diff
@@ -186,6 +186,6 @@ public override string ToString()
 
     internal abstract class SafeFreeContextBufferChannelBinding : ChannelBinding
     {
-        // TODO: To be implemented
+        // TODO (Issue #3362) To be implemented
```
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs]: You have an awful lot of TODO 3362, but 3362 only talks about SSL_CTX_set_quiet_shutdown.

Comment (rajansingh10) [inline src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs]: will update all those TODO in 3362.

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs (hunk 1 above)
Comment (rajansingh10) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: done

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs (hunk 2 above)
Comment (rajansingh10) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: done

**Comment by stephentoub** [comment]
@dotnet-bot test this please (Jenkins failure due to locked directory)

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs [hunk 4]
```diff
@@ -53,24 +48,26 @@ internal static IntPtr AllocateSslContext(long options, SafeX509Handle certHandl
...
                     throw CreateSslException("Failed to create SSSL object from SSL context");
-                }      
+                }
+
+                IntPtr memMethod = libcrypto.BIO_s_mem();
 
-                IntPtr memMethodRead = libcrypto.BIO_s_mem();
-                IntPtr memMethodWrite = libcrypto.BIO_s_mem();
-                if ((IntPtr.Zero == memMethodWrite) || (IntPtr.Zero == memMethodRead))
+                if (IntPtr.Zero == memMethod)
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: Reminder: especially now that the exceptions aren't being eaten by the catch block, the strings passed to these exceptions need to be in the .resx file so they can be localized.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs [hunk 5]
```diff
@@ -87,10 +84,11 @@ internal static IntPtr AllocateSslContext(long options, SafeX509Handle certHandl
...
             catch
             {
-                Marshal.StructureToPtr(sslContext, sslContextPtr, false);
-                FreeSslContext(sslContextPtr);
-                return IntPtr.Zero;
+                Disconnect(sslContext.sslPtr);
+                throw;
             }
+
+            IntPtr sslContextPtr = Marshal.AllocHGlobal(Marshal.SizeOf<SslContext>());
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: I'd previously provided feedback on this usage of SslContext as a struct, recommending instead that it be a class and that we just use a GCHandle rather than allocating native memory and marshaling the contents back and forth every time we need it.  Is that going to be done separately?

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs [hunk 6]
```diff
@@ -113,12 +111,14 @@ internal static bool DoSslHandshake(IntPtr sslContextPtr, IntPtr recvPtr, int re
...
+
             if (retVal != 1)
             {
-                error = GetSslError(context.sslPtr, retVal, "SSL Handshake failed: ");
+                error = GetSslError(context.sslPtr, retVal);
+
                 if ((retVal != -1) || (error != libssl.SslErrorCode.SSL_ERROR_WANT_READ))
                 {
-                    throw CreateSslException(context.sslPtr, "SSL Handshake failed: ", error);
+                    throw CreateSslException(context.sslPtr, "SSL Handshake failed: ", retVal);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: This is another example where this will need to be in the .resx file.  Most likely the string would end up being something like "SSL handshake failed with error {0}", and then the call site would use SR.Format to generate the string to pass to the exception.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs [hunk 7]
```diff
@@ -278,6 +248,19 @@ internal static void FreeSslContext(IntPtr sslContextPtr)
...
+        private static void FreeBio(SslContext sslContext)
+        {
+            if (IntPtr.Zero != sslContext.readBioPtr)
+            {
+                Interop.libcrypto.BIO_free(sslContext.readBioPtr);
+            }
+
+            if (IntPtr.Zero != sslContext.writeBioPtr)
+            {
+                Interop.libcrypto.BIO_free(sslContext.writeBioPtr);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: This is an example of where having the SslContext as a class instead of a struct would help.  Here you could set readBioPtr and writeBioPtr to null, preventing future accidental double-frees.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs [hunk 8]
```diff
@@ -414,16 +404,22 @@ private static SslException CreateSslException(IntPtr sslPtr, string message, in
 
         private sealed class SslException : Exception
         {
-            public SslException(string message)
-                : base(message + ": " + Marshal.PtrToStringAnsi(libssl.ERR_reason_error_string(libssl.ERR_get_error())))
+            private string message;
+            public override string Message { get { return message; } }
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: Why store the message like this?  How about just having a helper that produces the message given the ctor arguments and passes the result of that down to the base class?

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs [hunk 9]
```diff
@@ -414,16 +404,22 @@ private static SslException CreateSslException(IntPtr sslPtr, string message, in
...
+                this.message = inputMessage + ": " + Marshal.PtrToStringAnsi(libssl.ERR_reason_error_string(errorVal));
             }
 
-            public SslException(string message, int error)
-                : base(message + ": " + error)
+            public SslException(string inputMessage, int error)
+                : base()
             {
                 HResult = error;
+                this.message = inputMessage + ": " + error;
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: These string concats will need to be moved into the .resx file as strings with placeholders like {0} in them.

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs [hunk 10]
```diff
@@ -27,7 +27,15 @@ public void VerifyPackageInfo()
         public SafeFreeCredentials AcquireCredentialsHandle(X509Certificate certificate,
             SslProtocols protocols, EncryptionPolicy policy, bool isServer)
         {
-            return new SafeFreeCredentials(certificate, protocols, policy);
+            SafeFreeCredentials retVal = new SafeFreeCredentials(certificate, protocols, policy);
+            if ((null != retVal) && !retVal.IsInvalid)
+            {
+                // Caller does a ref count decrement
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: Immediately?  Otherwise this is at real risk of being leaked.

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs [hunk 11]
```diff
@@ -201,10 +209,15 @@ private long GetOptions(SslProtocols protocols)
...
-                    inputHandle = GCHandle.Alloc(inputBuffer.token, GCHandleType.Pinned);
-                    inputPtr = Marshal.UnsafeAddrOfPinnedArrayElement(inputBuffer.token, inputBuffer.offset);
-                    done = Interop.OpenSsl.DoSslHandshake(context.DangerousGetHandle(), inputPtr, inputBuffer.size, out outputPtr, out outputSize);
-                }           
+                    unsafe
+                    {
+                        fixed (byte* tokenPtr = inputBuffer.token)
+                        {
+                            inputPtr = new IntPtr(tokenPtr + inputBuffer.offset);
+                            done = Interop.OpenSsl.DoSslHandshake(context.DangerousGetHandle(), inputPtr, inputBuffer.size, out outputPtr, out outputSize);
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: What's keeping ```context``` alive during this call?

File: src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs (hunk 3 above)
Comment (rajansingh10) [inline src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs]: updated

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs (hunk 11 above)
Comment (rajansingh10) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: context.DangerousAddRef have been done earlier to keep it alive.

File: src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs (hunk 10 above)
Comment (rajansingh10) [inline src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs]: Added in #3362, will get it fixed.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs (hunk 4 above)
Comment (rajansingh10) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: This is a planned item, being tracked via #3362 . All .resx update will be done for all across.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs (hunk 5 above)
Comment (rajansingh10) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: yes, that will be done separately. This PR address some of the comments, not all.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs (hunk 6 above)
Comment (rajansingh10) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: Ok thanks for the detail in this regards, will be done as a separate PR for all .resx update.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs (hunk 8 above)
Comment (rajansingh10) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: done.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs (hunk 9 above)
Comment (rajansingh10) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: ok

**Comment by rajansingh10** [comment]
@stephentoub @bartonjs replied/fixed comments.

File: src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs (hunk 7 above)
Comment (rajansingh10) [inline src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs]: @vijaykota fyi pls
